dojo_plugin/pages/canvas.py

```diff
@@ -0,0 +1,225 @@
+        if "canvas_assignment_id" not in assessment:
+            continue
+        assessment_student_counter[assessment['id']] = 0
+        # if we have a value module_id, check for it and skip all the others
+        if module_id is not None and module_id != assessment["id"]:                
+            continue 
+                    
+        canvas_assignment_id = assessment["canvas_assignment_id"]
+        for grade_res in grades:                
+            grade_credit_percent = f"{grade_res['assessment_grades'][aindex]['credit'] * 100:.2f}%"
+
+            # this is a single module sync, submit to canvas and return
+            if module_id is not None and module_id == assessment["id"]:   
+                
+                res = post_grade_to_canvas(canvas_token, canvas_api_host, canvas_course_id, assessment["canvas_assignment_id"], 
+                                        students[grade_res["user_id"]], grade_credit_percent)
+                
+                return res
+            # multi-module and mutli-user sync of grades with canvas using bulk update api
+            else:
+                # additional check to be sure that they align 
+                if grade_res['assessment_grades'][aindex]['module_id'] == assessment['id']:
+                    student_id = students[grade_res['user_id']]
+                    canvas_grade_data[f"sis_user_id:{student_id}"] = {"posted_grade": grade_credit_percent}
+                    assessment_student_counter[assessment['id']] += 1
+                else:
+                    log.error(f"The following did not align in list, {grade_res['assessment_grades'][aindex]['module_id']=}   {assessment['id']=}" )
+                    return {"status": "failed", "message": f"The following did not align in list, {grade_res['assessment_grades'][aindex]['module_id']=}   {assessment['id']=}, this shouldn't happen."}
+
+        if (len(canvas_grade_data) > 0):
+            submission_results = post_bulk_grade_data_to_canvas(canvas_token, canvas_api_host, canvas_course_id, canvas_assignment_id, canvas_grade_data)
+            if submission_results["status"] == "success":
+                progress_urls.append(submission_results["url"])
+            else:
+                posting_results.append(submission_results)            
+        
+    if len(progress_urls) > 0 :
+        check_results = check_progress(progress_urls, canvas_token, interval=2)
+        posting_results.extend(check_results)
+        posting_results.append(assessment_student_counter)
+
+    return posting_results
+
+
+def post_grade_to_canvas(canvas_token, canvas_api_host, canvas_course_id, assignment_id, student_id, students_grade):
+    url = f"https://{canvas_api_host}/api/v1/courses/{canvas_course_id}/assignments/{assignment_id}/submissions/sis_user_id:{student_id}"
+    payload = {'submission': {'posted_grade': students_grade}}
+    
+    header = {'Authorization': 'Bearer ' + canvas_token}
```

Comment:
I'm not sold, but maybe an f-string here rather than string concatenation would be more clear (for `Bearer`)

Reply:
I'm not sure, they are both clear to me, but I'll change them all to be f-strings.